source/common/upstream/maglev_lb.cc

```diff
@@ -26,7 +26,8 @@ MaglevTable::MaglevTable(const NormalizedHostWeightVector& normalized_host_weigh
   table_build_entries.reserve(normalized_host_weights.size());
   for (const auto& host_weight : normalized_host_weights) {
     const auto& host = host_weight.first;
-    const std::string& address = host->address()->asString();
+    const std::string& address =
+        use_logical_name ? host->address()->logicalName() : host->address()->asString();
```

Comment:
Don't you mean to use `host->hostname()` here? AFAICT `logicalName()` is the same as `asString()`? I don't recall why we have `logicalName()` at all, I think maybe for customer resolvers? Do you use a custom resolver?

I think you intend for this to be used for DNS name if I remember the original issue? If so I think we probably need to change the API name per @htuch comment and make it a lot more clear what happens if there is no DNS name backing the host. We might actually want to fail the option completely if not using a DNS cluster?

Reply:
Thanks @mattklein123  and @htuch! Like I mentioned I wasn't sure if I should be using `host->hostname()` or `host->address()->logicalName()`. Most of the implementations of `address()->logicalName()` seems to be the same as `address()->asString()` except a few tests inside DNS resolver code. But I still wasn't able to find out how a host is constructed and how the host name we provide in the `socketAddress` API is used. 

I am happy to switch it to use hostName instead if that is the correct approach. I am not sure how this will work in EDS and how best to fail if DNS is not used and this option is enabled. I guess we can check if `hostname()` is empty and then fallback to using `address->asString()`. But again, I'm not sure where exactly hostName is set in the normal flow.

Reply:
@inf-rno yes based on what I think you want to accomplish (use DNS name for hashing vs. the IP address), I think you will need to `hostname()`. You can see how hostname is populated for strict DNS clusters here (logical DNS does something similiar):
https://github.com/envoyproxy/envoy/blob/3876cda8533ccb88bbbc4da0ee52b08cb3085eb0/source/common/upstream/strict_dns_cluster.cc#L119

You could potentially block this option in clusters other than DNS clusters somewhere during cluster creation, or maybe add some type of cluster interface that would say whether a cluster populates host name or not?

Either way, I think you should first mock up some changes and make sure they do what you think they do in your test environment. Thank you!

/wait

Reply:
See `ClusterFactoryImplBase::create`. You could potentially have an attribute of the cluster specify whether it fills hostname and then fail the load if not set and the user asks for the config option?